File: production/frontend/src/stores/scf.ts
/**
 * stores/scf.ts — 供应链金融 (SCF) 引擎状态管理
 *
 * 职责:
 *   1. 管理 SC6/SC7/SC8/SC9/SC10 计算结果与列表
 *   2. 4 个场景预设的一键加载与填表回写
 *   3. SCF↔Reform 联动 (R10 完成 → SC1 画像刷新 → SC8 撮合重算)
 *
 * project_memory 硬约束:
 *   - SC6 定价明细透明可解释 (baseRate/riskPremium/collateralDiscount/finalRate)
 *   - SC7 风险扩散每跳衰减, 不无限放大
 *   - SC8 撮合置信度 < 0.4 的候选过滤掉
 *   - SC9 红/黄/绿三色预警, 红灯必须银行介入
 *   - SC10 案例库四维分类 (行业/产品/规模/结果) + 关键词匹配
 */

import { defineStore } from 'pinia';
import { ref } from 'vue';
import * as scfApi from '@/api/scf';
import type {
  AlertsResult, CaseQuery, CaseRecord, MatchOutput, PricingOutput,
  ReformSyncEvent, ReformSyncResult, RiskPropagationOutput, SCFScenario,
  ScenarioId,
} from '@/api/scf';

export const useScfStore = defineStore('scf', () => {
  // === 状态 ===
  const pricingResult = ref<PricingOutput | null>(null);
  const riskPropagationResult = ref<RiskPropagationOutput | null>(null);
  const matchResults = ref<MatchOutput | null>(null);
  const alerts = ref<AlertsResult | null>(null);
  const cases = ref<CaseRecord[]>([]);
  const scenarios = ref<SCFScenario[]>([]);
  const currentScenario = ref<SCFScenario | null>(null);
  const lastSyncResult = ref<ReformSyncResult | null>(null);
  const loading = ref(false);
  const error = ref<string | null>(null);

  // === 通用错误处理 ===
  function _handleError(op: string, e: unknown): void {
    error.value = e instanceof Error ? e.message : String(e);
    console.error(`[ScfStore] ${op} 失败:`, e);
  }

  // === 动作 ===

  /** SC6 加载定价计算 */
  async function loadPricing(input: scfApi.PricingInput): Promise<PricingOutput | null> {
    loading.value = true;
    error.value = null;
    try {
      const result = await scfApi.pricing(input);
      pricingResult.value = result;
      return result;
    } catch (e) {
      _handleError('SC6 定价', e);
      return null;
    } finally {
      loading.value = false;
    }
  }

  /** SC7 加载风险扩散 */
  async function loadRiskPropagation(input: scfApi.RiskPropagationInput): Promise<RiskPropagationOutput | null> {
    loading.value = true;
    error.value = null;
    try {
      const result = await scfApi.riskPropagation(input);
      riskPropagationResult.value = result;
      return result;
    } catch (e) {
      _handleError('SC7 风险扩散', e);
      return null;
    } finally {
      loading.value = false;
    }
  }

  /** SC8 加载撮合结果 (top-K 候选) */
  async function loadMatch(input: scfApi.MatchInput): Promise<MatchOutput | null> {
    loading.value = true;
    error.value = null;
    try {
      const result = await scfApi.match(input);
      matchResults.value = result;
      return result;
    } catch (e) {
      _handleError('SC8 撮合', e);
      return null;
    } finally {
      loading.value = false;
    }
  }

  /** SC9 加载履约监控告警 */
  async function loadAlerts(): Promise<AlertsResult | null> {
    loading.value = true;
    error.value = null;
    try {
      const result = await scfApi.listAlerts();
      alerts.value = result;
      return result;
    } catch (e) {
      _handleError('SC9 履约监控', e);
      return null;
    } finally {
      loading.value = false;
    }
  }

  /** SC10 加载案例库 (支持 行业/产品/规模/结果/关键词 筛选) */
  async function loadCases(query?: CaseQuery): Promise<CaseRecord[]> {
    loading.value = true;
    error.value = null;
    try {
      const result = await scfApi.listCases(query);
      cases.value = result;
      return result;
    } catch (e) {
      _handleError('SC10 案例库', e);
      return [];
    } finally {
      loading.value = false;
    }
  }

  /** SC10 沉淀新案例 */
  async function saveCase(record: CaseRecord): Promise<CaseRecord | null> {
    loading.value = true;
    error.value = null;
    try {
      const result = await scfApi.saveCase(record);
      // 追加到本地列表
      cases.value = [result, ...cases.value];
      return result;
    } catch (e) {
      _handleError('SC10 沉淀案例', e);
      return null;
    } finally {
      loading.value = false;
    }
  }

  /** 加载 4 个场景预设 */
  async function loadScenarios(): Promise<SCFScenario[]> {
    loading.value = true;
    error.value = null;
    try {
      const result = await scfApi.listScenarios();
      scenarios.value = result;
      return result;
    } catch (e) {
      _handleError('加载场景预设', e);
      return [];
    } finally {
      loading.value = false;
    }
  }

  /** 加载单个场景预设 (回写 currentScenario + 返回 prefill) */
  async function loadScenario(scenarioId: ScenarioId): Promise<SCFScenario | null> {
    loading.value = true;
    error.value = null;
    try {
      const result = await scfApi.loadScenario(scenarioId);
      currentScenario.value = result;
      return result;
    } catch (e) {
      _handleError('加载场景预设', e);
      return null;
    } finally {
      loading.value = false;
    }
  }

  /** SCF↔Reform 联动 (接收 reform 完成事件) */
  async function syncFromReform(event: ReformSyncEvent): Promise<ReformSyncResult | null> {
    loading.value = true;
    error.value = null;
    try {
      const result = await scfApi.syncFromReform(event);
      lastSyncResult.value = result;
      // 若联动成功且触发了 rematch, 可选触发本地 match 刷新
      return result;
    } catch (e) {
      _handleError('SCF↔Reform 联动', e);
      return null;
    } finally {
      loading.value = false;
    }
  }

  /** 清空所有计算结果 (切换场景时调用) */
  function reset(): void {
    pricingResult.value = null;
    riskPropagationResult.value = null;
    matchResults.value = null;
    error.value = null;
  }

  return {
    // state
    pricingResult,
    riskPropagationResult,
    matchResults,
    alerts,
    cases,
    scenarios,
    currentScenario,
    lastSyncResult,
    loading,
    error,
    // actions
    loadPricing,
    loadRiskPropagation,
    loadMatch,
    loadAlerts,
    loadCases,
    saveCase,
    loadScenarios,
    loadScenario,
    syncFromReform,
    reset,
  };
}, {
  persist: {
    key: 'fintrust-scf',
    storage: sessionStorage,
    paths: ['currentScenario'],
  },
});

File: production/backend/app/api/v1/scf.py
"""供应链金融 (SCF) API 路由 — Tab11 SCF 工作台.

端点 (prefix=/scf):
    POST   /scf/pricing                  SC6 综合定价
    POST   /scf/risk-propagation         SC7 风险扩散
    POST   /scf/match                    SC8 撮合 top-K
    GET    /scf/alerts                   SC9 履约监控告警列表
    GET    /scf/cases                    SC10 案例库列表 (支持 ?industry=&product=&scale=&result=&keyword=)
    POST   /scf/cases                    SC10 沉淀新案例
    GET    /scf/scenarios                列出 4 个场景预设
    POST   /scf/scenarios/{scenario_id}/load    加载场景预设
    POST   /scf/sync-from-reform         SCF-09 接收 reform 完成事件

project_memory 硬约束:
    - 全部用 make_ok 包装响应
    - 全部用 CurrentUser 依赖 (app.deps.CurrentUser)
    - response_model 用 ApiResult[...] 形式
"""

from __future__ import annotations

from fastapi import APIRouter, Query, status

from app.api.deps import make_ok
from app.deps import CurrentUser
from app.schemas.common import ApiResult
from app.schemas.scf import (
    CaseOutcome,
    CaseQuery,
    CaseRecord,
    MatchInput,
    MatchOutput,
    PricingInput,
    PricingOutput,
    ReformSyncEvent,
    ReformSyncResult,
    RiskPropagationInput,
    RiskPropagationOutput,
    ScenarioId,
    ScfIndustry,
    ScfProduct,
    SCFScenario,
)
from app.services.scf_service import scf_service

router = APIRouter(prefix="/scf", tags=["Tab11 供应链金融"])


# ============================================================================
# SC6 定价
# ============================================================================

@router.post("/pricing", response_model=ApiResult[PricingOutput], summary="SC6 综合定价计算")
async def pricing(payload: PricingInput, _user: CurrentUser):
    """根据信用分/行业/担保方式/期限计算综合利率 (LPR + 风险溢价 - 担保抵扣)."""
    result = await scf_service.pricing(payload)
    return make_ok(result)


# ============================================================================
# SC7 风险扩散
# ============================================================================

@router.post("/risk-propagation", response_model=ApiResult[RiskPropagationOutput], summary="SC7 风险扩散")
async def risk_propagation(payload: RiskPropagationInput, _user: CurrentUser):
    """给定核心企业, 向上下游 N 跳传播风险 (应收账款变坏账 + 预付款损失)."""
    result = await scf_service.propagate_risk(payload)
    return make_ok(result)


# ============================================================================
# SC8 撮合
# ============================================================================

@router.post("/match", response_model=ApiResult[MatchOutput], summary="SC8 双向撮合")
async def match(payload: MatchInput, _user: CurrentUser):
    """企业融资需求与银行资金供给双向匹配 (信用+期限+担保偏好), 返回 top-K 候选."""
    result = await scf_service.match(payload)
    return make_ok(result)


# ============================================================================
# SC9 履约监控
# ============================================================================

@router.get("/alerts", response_model=ApiResult[dict], summary="SC9 履约监控告警列表")
async def list_alerts(_user: CurrentUser):
    """红黄绿灯告警列表 + 汇总统计."""
    alerts, summary = await scf_service.list_alerts()
    return make_ok({
        "alerts": [a.model_dump(by_alias=True) for a in alerts],
        "summary": summary.model_dump(by_alias=True),
    })


# ============================================================================
# SC10 案例库
# ============================================================================

@router.get("/cases", response_model=ApiResult[list[CaseRecord]], summary="SC10 案例库列表")
async def list_cases(
    _user: CurrentUser,
    industry: ScfIndustry | None = Query(default=None, description="行业筛选"),
    product: ScfProduct | None = Query(default=None, description="产品筛选"),
    scale: str | None = Query(default=None, description="规模筛选: micro/small/medium/large"),
    result: CaseOutcome | None = Query(default=None, alias="result", description="结果筛选: success/failed/partial"),
    keyword: str | None = Query(default=None, description="关键词模糊匹配 summary/tags"),
):
    """按 行业/产品/规模/结果 + 关键词 检索案例库."""
    query = CaseQuery(
        industry=industry, product=product,
        scale=scale,  # type: ignore[arg-type]
        outcome=result, keyword=keyword,
    )
    cases = await scf_service.list_cases(query)
    return make_ok([c.model_dump(by_alias=True) for c in cases])


@router.post("/cases", response_model=ApiResult[CaseRecord], status_code=status.HTTP_201_CREATED, summary="SC10 沉淀新案例")
async def save_case(payload: CaseRecord, _user: CurrentUser):
    """沉淀成功/失败案例到案例库."""
    result = await scf_service.save_case(payload)
    return make_ok(result)


# ============================================================================
# 场景预设 (SCF-08)
# ============================================================================

@router.get("/scenarios", response_model=ApiResult[list[SCFScenario]], summary="列出场景预设")
async def list_scenarios(_user: CurrentUser):
    """4 个场景预设: 反向保理 / 存货质押 / 应收账款转让 / 票据贴现."""
    scenarios = await scf_service.list_scenarios()
    return make_ok([s.model_dump(by_alias=True) for s in scenarios])


@router.post("/scenarios/{scenario_id}/load", response_model=ApiResult[SCFScenario], summary="加载场景预设")
async def load_scenario(scenario_id: ScenarioId, _user: CurrentUser):
    """加载场景预设, 返回 prefill 字段供前端自动填表."""
    result = await scf_service.load_scenario(scenario_id)
    if not result:
        return make_ok(None, code=404, message=f"场景预设 {scenario_id} 不存在")
    return make_ok(result.model_dump(by_alias=True))


# ============================================================================
# SCF ↔ Reform 联动 (SCF-09)
# ============================================================================

@router.post("/sync-from-reform", response_model=ApiResult[ReformSyncResult], summary="接收 reform 完成事件")
async def sync_from_reform(payload: ReformSyncEvent, _user: CurrentUser):
    """接收 reform R10 完成事件, 触发 SC1 画像刷新 + SC8 撮合重算."""
    try:
        result = await scf_service.sync_from_reform(payload)
        return make_ok(result)
    except Exception as e:
        return make_ok(None, code=-1, message=f"联动失败: {e}")
